417 reorder table column table

dstopp opened the pull request:
- Fixes #417 
- Were any Django, import pipeline, table_schema, or dictionary files modified? N
  - If YES:
    - Database used: opus3_test_200729
    - All Django tests pass: Y/NA
    - FLAKE8 run on modified code: Y/NA
- Were any JavaScript or CSS files modified? Y
  - If YES:
    - JSHINT run on all affected files: Y
    - Tested on Chrome, Firefox, Safari, Edge, and iOS: <still testing>
      (Remember to test all browser sizes)
    - Tested for race conditions (with delayed API calls if applicable): <still testing>

Description of changes:
 - added a dropdown toolbar to the table header that allows users to add either before or after the metadata element a new metadata field OR remove the selected metadata field.  

Known problems:
 - code for the reorder of the table headers is still in place but commented out

dstopp committed:
-allow user to sort via the table header; dim the columns during sort, move the columns so it looks good, then update the table w/a fetch of new data

dstopp committed:
change opacity

dstopp committed:
Merge branch 'master' of https://github.com/SETI/pds-opus into 417_reorder_table_column_TABLE

# Conflicts:
#	opus/application/static_media/js/browse.js

dstopp committed:
Merge branch 'master' of https://github.com/SETI/pds-opus into 417_reorder_table_column_TABLE

dstopp committed:
-initial checking that includes mouseover menu to allow adding or removing metadata fields from the table view interface
-reordering of table headers - not yet fabulous but it works

dstopp committed:
-fix the tooltips on the toolbar
-close any remnant toolbar or metadata field windows on resize

dstopp committed:
close the tools dropdown when the mouse leaves the table elements

dstopp committed:
-update tooltips

dstopp committed:
more tooltip updates

dstopp committed:
fix offset issue for the tools when the horizontal bar is moved

dstopp committed:
-fix for closing the metadata tools element

dstopp committed:
-smoother sorting of table headers

dstopp committed:
Merge branch 'master' of https://github.com/SETI/pds-opus into 417_reorder_table_column_TABLE

juzen2003 committed:
Modify .op-data-table sortable helper function to return a clone so that
the extra width will disappear when an item is grabbed.

juzen2003 committed:
Add the following 2 fixes:
1. when a header being grabbed, the text will stay at the bottom, it
   will not jump up.
2. when all header texts are single line and an item is grabbed, there
   will be no 2px right shift (2px difference comes from width difference
   between helper and placeholder).

dstopp committed:
-remove unused code
-remove/add in the 15px from the right side of the table header (there to account for PS issues)
-adjust the placeholder size

dstopp committed:
Merge branch '417_reorder_table_column_TABLE' into 417_reorder_table_column_TABLE

dstopp committed:
Merge pull request #11 from juzen2003/417_reorder_table_column_TABLE

Modify .op-data-table sortable helper function to return a clone so that

dstopp committed:
disable 'drag to reorder' for table headers

dstopp committed:
Merge branch '417_reorder_table_column_TABLE' of https://github.com/dstopp/pds-opus into 417_reorder_table_column_TABLE

# Conflicts:
#	opus/application/static_media/js/browse.js

dstopp committed:
-remove support for draggable table headers

dstopp committed:
-fix a typo; fix error in disabling button when there are no more metadata fields left to select

dstopp committed:
-make the toolbar font small when the screen is small

rfrenchseti commented on opus/application/static_media/js/utils.js:
```diff
@@ -154,3 +154,8 @@ $.fn.isOnScreen = function(scope, slop) {
     return (elementTop + offset <= bottom) && (elementTop >= top);
 };
+
+// more convienent this way... 
```
Spell convenient

rfrenchseti commented on opus/application/apps/ui/templates/ui/base.html:
```diff
@@ -265,11 +265,18 @@ <h3 class="op-header-text"></h3>
 
         <!-- dropdown menu for observations -->
-        <div class="dropdown-menu dropdown-menu-sm" id="op-obs-menu" data-toggle="dropdown" aria-labelledby="op-obs-menu" aria-haspopup="true" aria-expanded="false">
+        <div id="op-obs-menu" class="dropdown-menu dropdown-menu-sm"data-toggle="dropdown" aria-labelledby="op-obs-menu" aria-haspopup="true" aria-expanded="false">
```
Missing space before data-toggle

rfrenchseti commented on opus/application/apps/ui/templates/ui/base.html:
```diff
@@ -281,8 +288,14 @@ <h2 class="dropdown-header"></h2>
             <a class="dropdown-item" data-action="downloadURL" href="#" download><i class="fas fa-download"></i>Download zipped URL archive&nbsp;</a>
         </div>
 
+        <!-- drodown menu for adding metadata fields -->
```
Spell dropdown

rfrenchseti commented on opus/application/static_media/css/opus.css:
```diff
@@ -227,7 +227,7 @@ so that both "Recent Announcements" and "Message" can be vertically aligned. */
 }
 
-.op-metadata-detail-add {
+.#galleryView .op-metadata-detail-add {
```
Is there an extra "." before #galleryView?

rfrenchseti commented on opus/application/static_media/css/opus.css:
```diff
@@ -741,6 +756,22 @@ th.sticky-header {
 
+/* Fine tune the sortable helper text to stay at bottom in table header.
+Note: the original text moves up because the helper's default position is
+absolute, and this will result its child element ui-resizable (position: relative)
```
Bad grammar

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -397,6 +402,57 @@ var o_browse = {
+        $("#op-edit-field-tool a").on("click", function(e) {
+            let action = $(this).data("action");
+            let slug = $("#op-edit-field-tool").data("slug");
+            switch(action) {
```
Space after switch

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -1574,6 +1633,61 @@ var o_browse = {
+    initDraggableColumn: function(tab) {
+        function moveColumn(table, from, to) {
+            let rows = $('tr', table);
+            let cols;
```
Why is cols set outside of the loop?

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -1574,6 +1633,61 @@ var o_browse = {
+            },
+            stop: function(e, ui) {
+                o_browse.isSortingHappening = false;
+                let columnOrder = $(this).sortable( "toArray" );
```
Extra spaces

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -1574,6 +1633,61 @@ var o_browse = {
+
+                // only bother if something actually changed...
+                if (!o_utils.areObjectsEqual(opus.prefs.cols, columnOrder)) {
+
```
Extra blank line

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -2210,30 +2326,83 @@ var o_browse = {
+        o_utils.disableUserInteraction();
+
+        o_hash.updateURLFromCurrentHash(); // This makes the changes visible to the user
+        // passing in false indicates to not close the gallery view on loadData
```
Passing in false to what?

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -2210,30 +2326,83 @@ var o_browse = {
+
+    onDoneUpdateFromTableMetadataDetails: function(e) {
+        o_utils.disableUserInteraction();
+
```
Extra blank line

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -2210,30 +2326,83 @@ var o_browse = {
-        let contextMenu = "#op-add-metadata-fields";
-        let targetPosition = $(e.currentTarget).position();
-        let left = targetPosition.left;
+        let targetPosition = $(e.currentTarget).offset();
```
Something in here seems to have changed. With a little work I can get the *slideshow* add-metadata dropdown to dropup when it shouldn't, and run a little off the top of the browser. If I have exactly the same screen on the production OPUS, it doesn't do it.

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -2293,6 +2462,7 @@ var o_browse = {
     },
 
     checkForEmptyMetadataList: function() {
+        // if there is only one metadata field left, disable the trash in both the gallery view and table tools dropdown
```
Gallery view? Do you mean slideshow?

rfrenchseti reviewed (changes requested):
- I haven't found a way to reproduce it with 100% reliability, but if you add and delete columns a lot you eventually get to a state where the "add/delete" toolbar ends up on top of the header, instead of below. Once it's there, it never gets better. When this happens, it's also slightly offset from the header position - a few pixels to the right.

dstopp committed:
-address comments in pull request

dstopp committed:
-fixed issue where the toolbar would move down over time and become offset
-fixed issue where user interaction was not disabled on add of a new field from the dropdown tool <delete was disabled, only add was not>

dstopp committed:
-remove unnecessary commented out code

dstopp committed:
-close the toolbar when the mouse moves to the left and down below <can be tested by moving the mouse fast at a diagonal>

dstopp committed:
-close toolbar when the mouse moves fast vertically straight down
